### ISSUE TITLE
on altIdent

### ISSUE THREAD

**Issue by sydb** [TEIC/TEI#2049]
(Note: This issue wraps in the problem initially posted in #1859, the problem later discovered in #1859, and one of the last comments of #1919, namely that “there are some TD elements that probably should not have `<altIdent>`, either; `<modelGrp>`, e.g.”.)

# Problems

The problems with `<altIdent>` are:
* It is still allowed in some places where perhaps it should not be.
* Its content model is silly: it should be `<dataRef key="teidata.xmlName"/>`.
* If the content model is changed, it could no longer be used is as a child of `<moduleSpec>` (with `@type` of "FPI"). Which would be a good thing, as its use there is simply incorrect — it does not match the definition of `<altIdent>` at all. But we need to find another way to represent the FPI of each module (which is needed for DTD support). Towards the end of this comment please find a table of the identifier of each module along with the heading of its chapter and its FPI.

# Plan

So my suggested plan of action is, roughly, as follows.

1. Decide on an alternate encoding of FPIs. My current favorite is to use an `<idno type="FPI">` exactly where we currently use `<altIdent type="FPI">`.
2. Decide which elements should, and should not, have `<altIdent>` in their content model.
3. Write Schematron constraint to manually deprecate use of `<altIdent>` in places identified in step 2 as no-nos.
1. Write another Schematron constraint to ensure that all `<altIdent>`s are xsd:NCName (the equivalent of teidata.xmlName) EXCEPT those that are a child of `<moduleSpec>` and (or or?) have a `@type` of "FPI".
4. Wait 6 months to a year.
5. During this time, if even 1 person asks for it, write a stylesheet to convert `moduleSpec/altIdent` to whatever we decided in step 1.
6. During this time modify all Stylesheets so that they can read and process both the method agreed in step 1 and the current method for finding the FPI. (That is, either will work. There are 62 occurrences of "altIdent\b" in 10 Stylesheet files, although a few of those are comments.)
1. Modify content models, classes, etc., so that `<altIdent>` is not allowed in the silly places identified in step 2.
1. Remove manual deprecation Schematron.
1. Modify content model of `<altIdent>` itself (to `<dataRef key="teidata.xmlName" minOccurs="1" maxOccurs="1"/>`)
1. Remove constrain-to-xsd:NCName Schematron.
1. Change our own use of `moduleSpec/altIdent` to whatever we decided in step 1.
1. Wait another 6 months to a year.
1. Remove the Stylesheets’ capability to handle the old `<altIdent>` method for indicating the FPI of a module.

So, presuming Council agrees with this overall plan, the main issues that would need discussion and decision are discussed in the next 2 sections.

# Where should `<altIdent>` be allowed?

The list of elements where it is currently allowed is: `<attDef>`, `<classSpec>`, `<constraintSpec>`, `<dataSpec>`, `<elementSpec>`, `<macroSpec>`, `<model>`, `<modelGrp>`, `<modelSequence>`, `<moduleSpec>`, `<paramSpec>`, `<schemaSpec>`, and `<valItem>`.

It is clear (to me, at least) that it is required for use in `<elementSpec>` and `<attDef>`, and I believe that the Stylesheets will do the right thing in those cases (although I have not recently tested). I do not know what the Stylesheets do for `valItem/altIdent`, but I can certainly imagine uses. At the moment (way past my bedtime), I am having trouble imagining what `<altIdent>` as a child of any of the others would mean.

The basic idea, I think, is that use of `<altIdent>` allows you to separate the name as used in the ODD (e.g., as the value of `@key`) and RELAX NG schema (as the pattern name) from the name of the construct in the schema and instances. Is that ever helpful for `<dataSpec>` or `<paramSpec>`?

# How should we encode the FPI of a module?

As mentioned above, my instinct is to use `<idno type="FPI">`. Other possibilities include:
* `<label>`
* `@n`
* a new attribute, e.g. `@fpi`
* `<head>`
* `<ident>`
* `<name>`
* a new element, e.g. `<fpi>`
* `<gloss>`
* `<rs>`

# Addendum

| `@ident` | heading | FPI |
| --- | ------------------ | ----------------------------- |
|tei | The TEI Infrastructure | TEI Infrastructure |
|header | The TEI Header | Common Metadata |
|core | Elements Available in All TEI Documents | Common Core |
|textstructure | Default Text Structure | Default Text Structure |
|gaiji | Characters, Glyphs, and Writing Modes | Character and Glyph Documentation |
|verse | Verse | Verse |
|drama | Performance Texts | Performance Texts |
|spoken | Transcriptions of Speech | Transcribed Speech |
|dictionaries | Dictionaries | Dictionaries |
|msdescription | Manuscript Description | Manuscript Description |
|transcr | Representation of Primary Sources | Transcription of Primary Sources |
|textcrit | Critical Apparatus | Text Criticism |
|namesdates | Names, Dates, People, and Places | Names, dates, persons and places |
|figures | Tables, Formulæ, Graphics and Notated Music | Tables, Formulæ, Notated Music, Figures |
|corpus | Language Corpora | Metadata for Language Corpora |
|linking | Linking, Segmentation, and Alignment | Linking, Segmentation, and Alignment |
|analysis | Simple Analytic Mechanisms | Analysis and Interpretation |
|iso-fs | Feature Structures | Feature Structures |
|nets | Graphs, Networks, and Trees | Graphs, networks, and trees |
|certainty | Certainty, Precision, and Responsibility | Certainty and Uncertainty |
|tagdocs | Documentation Elements | Documentation Elements |

**Comment by sydb** [comment]
Instructions for Council subgroup (@sydb, @ebeshero, @npcole, and special guest @martindholmes):

Fill in “1” (favorite), “2” (second favorite), “3” (third favorite) in your three favorite methods for encoding what we currently encode as `<altIdent type="FPI">` in the column under your initials.

Instructions for anyone else who would like to express an opinion:

Edit the table adding a column for yourself, and let us know your preferred method(s) using the same numbering system.

| encoding                                             | SB | EBB | NC | MH |
| ---------------------------------------------------- | -- | --- | -- | -- |
| a new element, e.g. `<fpi>`                          |    |     |    |    |
| a new attribute, e.g. `@fpi` (on the `<moduleSpec>`) |    |     |    |    |
| `<gloss>`                                            |    |     |    |    |
| `<gloss type="FPI">`                                 |    |     |    |    |
| `<head type="FPI">`                                  |    |     |    |    |
| `<ident>`                                            |    |     |    |    |
| `<idno type="FPI">`                                  | 1  | 1    |    |  1  |
| `<label type="FPI">`                                 | 2  |  3   |    | 2  |
| `@n` (on the `<moduleSpec>`)                         | 3  |  2   |    |  3  |
| `<name>`                                             |    |     |    |    |
| `<rs>`                                               |    |     |    |    |
| _Council has already expressed distaste_ | _for_ | _those_ | _below_ | _here_ |
| `<ident type="FPI">`                                 |    |     |    |    |
| `<name type="FPI">`                                  |    |     |    |    |
| `<rs type="FPI">`                                    |    |     |    |    |

**Comment by sydb** [comment]
Instructions for Council subgroup (@sydb, @ebeshero, @npcole, and special guests @jamescummings and @martindholmes):

In the column under your initials fill in a checkmark (‘✓’, U+2713) for those elements for which you think `<altIdent>` should be allowed as a child; fill in a ballot X (‘✗’, U+2717) for those elements for which you think `<altIdent>` should **not** be allowed as a child; fill in a question mark for those for which you think more discussion is required; and leave blank those for which you are ambivalent, uncertain, or just don’t give a hoot. If there is an element which you think _should_ permit a child `<altIdent>` but is not in the table (because currently it is neither allowed an `<altIdent>` child nor has an `@ident` attribute), please add a new row for it to the end of the table.

Instructions for anyone else who would like to express an opinion:

Edit the table adding a column for yourself, and let us know where you think `<altIdent>` should and should not be allowed using the same system.

Explanation of columns:

1. **parent**: name of element that should or should not have a child `<altIdent>`; if in italics this element is not in the tagdocs module, and thus an `<altIdent>` child is or would be ill-defined (unless definition of `<altIdent>` is updated).
1. **`@ident`**: indicates if this element has an `@ident` attribute via the att.identified class, as a direct attribute defined for this element, or not at all.
1. **child `<altIdent>`**: indicates if this element currently allows a child `<altIdent>` or not; if it does, it is via the model.glossLike class.

| parent                   | has<br/> `@ident`?  | has<br/>child<br/>`<altIdent>`<br/>? | SB | EBB | NC | JC | MH |
| ------------------------ | -------------- | ----------------------- | -- | --- | -- | -- | -- |
| _application_    	   | direct 	    | NO 		      | ?  |   X  |    |   ✗ |?    |
| attDef    		   | att.identified | yes		      | ✓  |  ✓   |    |   ✓ | ✓ |
| _category_    	   | NO 	    | yes		      | ✗  |  ✓   |    |    ✗| X   |
| classSpec    		   | att.identified | yes		      | ✓  |   ✓  |    |   ✓ | ✓  |
| constraintSpec    	   | att.identified | yes		      | ✓  |   ✓  |    |   ✓ |  ✓  |
| dataSpec    		   | att.identified | yes		      | ✓  |   ✓  |    |    ✓|  ✓  |
| elementSpec    	   | att.identified | yes		      | ✓  |  ✓   |    |   ✓ |  ✓  |
| _joinGrp_    		   | NO 	    | yes		      | ✗  |   X  |    |   ✗ |  X |
| _language_    	   | direct 	    | NO		      | ✗  |   X  |    |   ✗ |  ✓  |
| macroSpec    		   | att.identified | yes		      | ✓  |  ✓   |    | ✓   |  ✓  |
| model    		   | NO 	    | yes		      | ?  | X    |    |    ✗|  X  |
| modelGrp    		   | NO 	    | yes		      | ?  |  X   |    |    ✗| X   |
| modelSequence    	   | NO 	    | yes		      | ?  |   X  |    |    ✗| X  |
| moduleSpec    	   | att.identified | yes		      | ✗  |   ?  |    |   ✓ |  ✓  |
| paramSpec    		   | att.identified | yes		      | ?  |   ✓  |    |  ✗ |  ✓ |
| _prefixDef_    	   | direct 	    | NO		      | ✗  |  X   |    |   ✗ |  X  |
| remarks    		   | direct 	    | NO		      | ✗  |  X   |    |   ✗ |  X  |
| schemaSpec    	   | att.identified | yes		      | ✗  |   ?  |    |?    |  ? |
| _taxonomy_    	   | NO 	    | yes		      | ✗  |  X   |    |    ✗|  X  |
| _transcriptionDesc_      | direct 	    | NO		      | ✗  |   X  |    |  ✓  |  X  |
| valItem    		   | direct 	    | yes		      | ✓  |    ✓ |    |    ✓|  ✓ |
| **new** 		   |  **places**    | **go**  | **be-** | **low** | **here** | (-: | :-) |

**Comment by martindholmes** [comment]
I'm sure this was covered in the Council discussion, but I wasn't there so I'm going to ask: Aren't we confusing a human-readable title with a Formal Public Identifier here? "Names and Dates" is a heading or a title; an FPI should presumably be some sort of unique identifier such as a URI, shouldn't it? We would expect "Names and Dates" to be translated into another language when the schema is not expressed in English, but would we want the FPI to change?

**Comment by sydb** [comment]
No, I don’t think we are confusing the title (e.g. “The TEI Header”, which is found at `./ancestor::div[not(ancestor::div)]/head` in p5.xml, where `.` is the `<moduleSpec>` in question) with the FPI (e.g. “-//TEI P5//ELEMENTS Common Metadata//EN”, of which only the “Common Metadata” bit is stored as encoded information in p5.xml, at `./altIdent[@type eq 'FPI']`). While an FPI provides a mechanism for translation into other languages (hence the ending “//EN”), I do not think I have ever seen it used. (The only place I expected to find one in TEI-land is in P5/Source/Guidelines/**fr**/HD-Header.xml, but it does not even have a `<moduleSpec>` in which to have a child `<altIdent type="FPI">`!)

HOWEVER, I believe I have just discovered that I have the “FPI” and “heading” columns reversed in the table in the Addendum of the initial post. I have a meeting in 1 min, but after that plan to check & fix if needed.

**Comment by sydb** [comment]
Worth mentioning there is an FPI → URI methodology:
~~~~
urn:publicid:-:TEI P5:ELEMENTS Common Metadata:EN
~~~~~

**Comment by sydb** [comment]
Yeah, I had the columns in the **Addendum** above reversed. Fixed now.

**Comment by sydb** [comment]
### Plan, `#2`: where should `<altIdent>` be allowed?

Instructions for Council and other interested parties:

Edit the table adding a column for yourself, and let us know where you think `<altIdent>` should and should not be allowed using the following system. (Note: it is _much_ easier to edit the table if you use an editor with a monospace or fixed-width font.)

In the column under your initials fill in a checkmark (‘✓’, U+2713) for those elements for which you think `<altIdent>` should be allowed as a child; fill in a ballot X (‘✗’, U+2717) for those elements for which you think `<altIdent>` should **not** be allowed as a child; fill in a question mark for those for which you think more discussion is required; and leave blank those for which you are ambivalent, uncertain, or just don’t give a hoot. If there is an element which you think _should_ permit a child `<altIdent>` but is not in the table (because currently it is neither allowed an `<altIdent>` child nor has an `@ident` attribute), please add a new row for it to the end of the table.

Note that this table only includes entries for those places about which the sub-group (SB, EB, JC, MH) had some internal disagreement. I have also removed `<moduleSpec>`, as we have agreed to replace `moduleSpec/altIdent[@type='FPI']` with `moduleSpec/ident[@type='FPI']` (see [above1](#issuecomment-725084222)). You can see the complete list [above2](#issuecomment-725088683).

Explanation of columns:

1. **parent**: name of element that should or should not have a child `<altIdent>`; if in _italics_ this element is not in the **tagdocs** module, and thus an `<altIdent>` child is or would be ill-defined (unless definition of `<altIdent>` is updated).
1. **`@ident`**: indicates if this element has an `@ident` attribute via the att.identified class, as a direct attribute defined for this element, or not at all.
1. **child `<atIdent>`**: indicates if this element currently allows a child `<altIdent>` or not; if it does, it is via the model.glossLike class.

| parent                   | has<br/> `@ident`?  | has<br/>child<br/>`<altIdent>`<br/>? | SB | EBB | JC | MH | MS | HC | RV | HBS |
| ------------------------ | -------------- | ----------------------- | -- | --- | -- | -- | -- | -- | -- | -- |
| _application_    	   | direct 	    | NO 		      | ?  |  ✗  | ✗  | ?  | ✗ | ✗ | ✗ | ✗ | 
| _category_    	   | NO 	    | NO		      | ✗  |  ✓  | ✗  | ✗  | ✗ | ✗ | ✗ | ✗ |
| _language_    	   | direct 	    | NO		      | ✗  |  ✗  | ✗  | ✓  | ✗ | ✓ | ✓ | ✓ |
| paramSpec    		   | att.identified | yes		      | ✗  |  ✓  | ✗  | ✓  | ? | ✓ | ✗ | ?  | 
| schemaSpec    	   | att.identified | yes		      | ✗  |  ?  | ?  | ?  | ✗ | ✗ | ✗ | ?  | 
| _transcriptionDesc_      | direct 	    | NO		      | ✗  |  ✗  | ✗  | ✗  | ✗ | ✗ | ✗ | ✗ |
| **new** 		   |  **places**    | **go**  | **be-** | **low** | **here** | (-: | :-) |

**Comment by sydb** [comment]
## summary of step (2) results so far

### uncontroversial
That is, subgroup agreed in 1st round of voting. 

#### per step (1)
`<moduleSpec>` loses (by deprecation) its `<altIdent>` child, to be replaced with `<idno>` in our ODDs

#### do _not_ already have `<altIdent>`, and stay that way
_`<prefixDef>`_, `<remarks>`, and _`<transcriptionDesc>`_

#### already do have `<altIdent>`, and stay that way
`<attDef>`, `<classSpec>`, `<constraintSpec>`, `<dataSpec>`, `<elementSpec>`, `<macroSpec>`, and `<valItem>`

#### lose their current child `<altIdent>`
`<joinGrp>`, `<model>`, `<modelGrp>`, `<modelSequence>`, and `<taxonomy>`

#### gain a child `<altIdent>`
[none]

### comparatively controversial
That is, went to 2nd round of voting.

#### do _not_ already have an `<altIdent>` child and are _not_ in the `tagdocs` module
* `<application>`
  * votes: 6 ✗, 2 ?, 0 ✓
  * result: no change (does not get a new `<altIdent>` child)
* `<category>`
  * votes: 7✗, 0 ?, 1 ✓
  * result: no change (does not get a new `<altIdent>` child; EBB is welcome to open a new ticket to argue that it should)
* `<language>`
  * votes: 4✗, 0 ?, 4 ✓
  * result: NEEDS DISCUSSION, although if no consensus can be reached I will make no change on this ticket; we can open a new ticket if desired
* `<transcriptionDesc>`
  * votes: 8✗, 0 ?, 0 ✓
  * result: no change (does not get a new `<altIdent>` child; JC has changed his mind.)

#### already have an `<altIdent>` child and are in the `tagdocs` module
* `<paramSpec>`
  * votes: 2 ✗,  3 ?, 3 ✓
  * result: NEEDS DISCUSSION, although if no consensus can be reached I will (begrudgingly) leave `<altIdent>` in the content model of `<paramSpec>`
* `<schemaSpec>`
  * votes: 4 ✗, 4 ?, 0 ✓
  * result: no change (does not get a new `<altIdent>` child)

**Comment by sydb** [comment]
So, there are 2 elements left to discuss for step (2), `<language>` and `<paramSpec>`.

### on `<language>`

(With apologies to William Safire.)

Left to my own devices, I am **very** much against adding `<altIdent>` to the content model of `<language>`. That said, it is quite possible that @martindholmes, @hcayless, @raffazizzi, or @HelenaSabel have arguments in favor that will convince me.

My first thought is “what would it _mean_?”. The value of `@ident` is “a language code constructed as defined in BCP 47 which is used to identify the language documented by this element, and which is referenced by the global xml:lang attribute”. But `<altIdent>` “supplies the recommended XML name for an element, class, attribute, etc. in some language” (not in some `<language>` :-) I am hard pressed to think of languages as part of the “etc.”, there.

Even if we were to change the _semantics_ of `<altIdent>` to allow it to be useful in the context of `<language>`, the _syntax_  would likely be too constraining. (And remember, a _large_ part of this ticket is about reinforcing and properly constraining the syntax of `<altIdent>`).

Besides, how would it be helpful to encode an alternate identifier? If it is just to say “Ellen encoded this unusual language with "x-kn" and Anne encoded it with "x-kr", so here we use both identifiers for one `<language>`” as a way of avoiding a duplicate `<language>`, I have no sympathy. If you are thinking it is the way to tuck in the equivalent ISO 639 (or whatever) code, again, I have no sympathy. (I would say that should be encoded with an `<idno>` with  `type="ISO639"` (or whatever), which is already allowed inside `<language>`.)

My second thought is that `<language>` is currently the only element being considered as a possible parent of `<altIdent>` that is not from the tagdocs module. That means that if we allow `<altIdent>` as a child of `<language>` we need to have a floppy and vague or dual-tracked definition. If we reject this notion, the description gets to remain precise.

### on `<paramSpec>`

Left to my own devices, I am moderately in favor of removing `<altIdent>` from the content model of `<paramSpec>`. That said, it is quite possible that @ebeshero , @martindholmes, or @hcayless have arguments against that will convince me.

I note that there is no occurrence of `paramSpec/altIdent` in the _Guidelines_, including in any example, nor is `<altIdent>` mentioned at all in [22.5.4 Processing Models](https://www.tei-c.org/release/doc/tei-p5-doc/en/html/TD.html#TDPM). I doubt anyone has any system that uses a `paramSpec/altIdentifier` or knows how to process one. (@jamescummings? @tuurma?)

I presume that an alternate identifier of a parameter means that the parameter could be invoked with either the main (`@ident`) or one of the alternate (`<altIdent>`) identifiers. Does anyone know any other system in which parameters can have multiple names? Well, yes, we all do. E.g.,
~~~sh
$ diff --text --ignore-space-change
~~~
is the same as
~~~sh
$ diff -a -b
~~~
Even for those who think this is a good thing, is it helpful enough to be worth adding this complexity?

**Comment by martindholmes** [comment]
On `<language>`, @sydb says: 

> I would say that [iso-639 identifiers etc.] should be encoded with <idno type="ISO639">, which is already allowed inside <altIdent>.

I presume that should be "allowed inside `<language>`.

Other than that, I think I find your argument convincing. My reason for voting for <altIdent> here was that I'm constantly aware that languages (especially endangered or extinct languages) are frequently described, identified and referred to in many different ways by different communities, but of course this really has no bearing on `<altIdent>` in its strict definition as a component of ODD schema specifications, so I hereby change my vote.

**Comment by sydb** [comment]
Yes, @martindholmes, the MarkDown source of that snippet says
~~~
(I would say that should be encoded with an `<idno>` with  `type="ISO639"` (or whatever), which is already allowed inside `<altIdent>`.)
~~~
so it seems very odd that you are not seeing the backticks. (Even your copied quote from my post has the tags in it without the backticks.)

Anyway, `<language>` now has 5 ✗ and 3 ✓. Anyone (calling @hcayless, @raffazizzi, and @HelenaSabel) want to argue strongly in favor of putting `<altIdent>` inside `<language>`?

**Comment by martindholmes** [comment]
@sydb I wasn't talking about the backticks; I was talking about the element name "altIdent", which I think is a typo for "language", You wanted to say (I think) that you can use idno inside language, so you don't need altIdent. Instead you implied that you can use idno inside altIdent, which isn't true.

**Comment by hcayless** [comment]
My motivation was identical to Martin's, as was my mistake in not realizing that `<altIdent>` only works in tandem with the `@ident` in `att.identified`. Not going to argue in favor :-). 

As for `<paramSpec>`, I'm afraid I don't have the bandwidth to explain the Processing Model to myself again, but wouldn't its use be analogous to that of other tagdoc elements? Namely, to provide an alternate name to use in some data structure derived from the ODD? Maybe that's not something you'd ever want to do though...

**Comment by sydb** [comment]
Right! Thank you, @martindholmes. I have just fixed that, which means that future readers of this exchange will be **quite** confused. :-)

**Comment by sydb** [comment]
OK, @hcayless, that makes it 6:2 against putting `<altIdent>` into `<language>`. So I think that settles it, I won’t be doing that on this ticket. If someone wants to create another special-purpose ticket for that, so be it.

As for `<paramSpec>` I think your analysis is correct, it provides an alternate name to use in the ODD to refer to the parameter. But I cannot (off the top of my head, which like yours is not really in processing-model-space) see a reason one would want to refer to the same parameter with more than 1 name. (I do not think the parameter gets expressed in RELAX NG or other outputs, does it?)

**Comment by sydb** [comment]
### more on `<paramSpec>`

Realizing that one of the ‘?’ votes was mine, I changed it to ✗, leaving us with a vote tally of 3 ✗, 2 ?, and 3 ✓. So IMHO it is sorta up to MH or HC to come up with a compelling reason as to _why_ `<altIdent>` should be left as a child of `<paramSpec>`.

**Comment by jamescummings** [comment]
re: `<paramSpec>` 

@sydb / @hcayless : I think that your understanding of `<paramSpec>` is correct; it would be an alternative ident for the parameter name. Of all the other strange places that we're getting rid of `<altIdent>` from, this is the one where it makes _most_ sense to remain. But I don't have a particular use-case for it. Having aliasing of parameters might be useful where `<paramSpec>` is being used in a different contexts where an existing vocabulary of parameters exists in a particular coding framework. That's the only use-case I can imagine to be honest. I've no strong feelings about keeping or removing it, but it makes more sense there (since it is a tagDocs Spec at least) than places like `<language>` or `<category>` or `<transcriptionDesc>`.

**Comment by lb42** [comment]
In general it seems perfectly plausible to have alternative forms of a parameter name. Look at the spec for any Unix command.

**Comment by martindholmes** [comment]
Also @sydb you did say before "if no consensus can be reached I will (begrudgingly) leave `<altIdent>` in the content model of `<paramSpec>`".

**Comment by sydb** [comment]
Yes, @lb42, it is completely plausible. And yes, @martindholmes, I did say that. But (with the possible exception of you) we have just reached consensus in that all of Council (including HC and EB, and also JC) have agreed to dropping it. We (including MT) believe, BTW, that there is no software in the world that would have any idea what to do with `paramSpec/altIdent`.

The Unix-command switch duality that both @lb42 and I brought up is (in my mind) a reasonable but misleading analogy. After all (JC pointed out), the `@ident` to which `<altIdent>` would provide an alternate is already going to be the long, human-readable version. Not much of a benefit in having a shorter name for something that is only used in your ODD.

We could all think of use cases; none of us could come up with any _compelling_ use cases.

**Comment by ebeshero** [comment]
@martindholmes @sydb Here is what persuaded me toward consensus on removing `<altIdent>` from `<paramSpec>`: I asked if there might be another way to indicate an alternative to a parameter. The answer seems to be yes, and on closer inspection of the content model we have, I think `<equiv>` would satisfy if needed.

**Comment by sydb** [comment]
Or just having a 2nd `<paramSpec>`, identical save the `@ident`.

**Comment by sydb** [comment]
### Where we are so far — Or, Plan Reduex

Here are the steps from the original post on this ticket, annotated with what has been done so far.

1. ~~Decide on an alternate encoding of FPIs. My current favorite is to use an `<idno type="FPI">` exactly where we currently use `<altIdent type="FPI">`.~~ — DONE, we have agreed to `<idno type="FPI">`
2. ~~Decide which elements should, and should not, have `<altIdent>` in their content model.~~ — DONE, `<altIdent>` is to be limited to being a child of `<attDef>`, `<classSpec>`, `<constraintSpec>`, `<dataSpec>`, `<elementSpec>`, `<macroSpec>`, or `<valItem>`;
3. ~~Write Schematron constraint to manually deprecate use of `<altIdent>` in places identified in step 2 as no-nos.~~ —DONE, #2193 deprecates them everywhere else, but see [1]
4. ~~Write another Schematron constraint to ensure that all `<altIdent>`s are xsd:NCName (the equivalent of teidata.xmlName) EXCEPT those that are a child of `<moduleSpec>` and (or or?) have a `@type` of `"FPI"`.~~ — DONE, but see [1]
5. Wait 6 months to a year. — Ongoing, as it were.
6. ~~During this time, if even 1 person asks for it, write a stylesheet to convert moduleSpec/altIdent to whatever we decided in step 1.~~ — DONE. One person did ask for it: me, to perform step 12. (I have not committed the code, yet, it is just on my local system as altIdent2idno4FPIs.xslt. It is pretty simple, with a totoal of 34 elements + attributes.)
7. During this time modify all Stylesheets so that they can read and process both the method agreed in step 1 and the current method for finding the FPI. (That is, either will work. There are 62 occurrences of "altIdent\b" in 10 Stylesheet files, although a few of those are comments.) — I _think_ this is DONE with #510.
8. Modify content models, classes, etc., so that `<altIdent>` is not allowed in the silly places identified in step 2.
1. Remove manual deprecation Schematron.
1. Modify content model of `<altIdent>` itself (to `<dataRef key="teidata.xmlName" minOccurs="1" maxOccurs="1"/>`)
1. Remove constrain-to-xsd:NCName Schematron.
1. ~~Change our own use of `moduleSpec/altIdent` to whatever we decided in step 1.~~ DONE in 792a126c4262a4ecb26f4323931907bfd23af6cb.
1. Wait another 6 months to a year.
1. Remove the Stylesheets’ capability to handle the old `<altIdent>` method for indicating the FPI of a module.

So I _think_ there is nothing else to do for this ticket now except to change the end-deprecation date to mid-October. Then, in mid-October we do steps 8, 9, 10, & 11. (Then, in 2023 we do step 14.)

#### Notes
[1] I set the valid-until date to 2022-07-14 back when I did this (2021-10-11), in order to have a date that was at least 6 months in the future and perhaps even 6 months after the next release. However, the PR was not merged until 2022-04-02, a mere ~3.2 months before the valid-until date. I think we should reset the valid-until date to 2022-10-21, and plan to just do so soon.

**Comment by sydb** [comment]
**TTD?**

Revisiting this after months of inattention.

> So I _think_ there is nothing else to do for this ticket now except to change the end-deprecation date to mid-October. Then, in mid-October we do steps 8, 9, 10, & 11. (Then, in 2023 we do step 14.)

Checking now:

- Step 8: done
- Step 9: done
- Step 10: BLOCKED by [S#569](https://github.com/TEIC/Stylesheets/issues/569)
- Step 11: done
- Step 13: what we are doing right now
- Step 14: should be done in late 2023 at the earliest, IMHO.

Therefore I think we should (right now):
1. Add a comment to [S#569](https://github.com/TEIC/Stylesheets/issues/569) reminding us to fix the content model of `<altIdent>` when the Stylesheets are fixed. — DONE
2. Close this ticket as (essentially) finished. — DONE
3. optional: Create a Stylesheets ticket to remind ourselves to “Remove the Stylesheets’ capability to handle the old `<altIdent>` method for indicating the FPI of a module.”. — DONE
